### ISSUE TITLE
clippy::use_self seems to trigger in some odd places

### ISSUE THREAD

**Issue by c0repwn3r** [rust-lang/rust-clippy#10127]
### Summary

I have a simple enum which for some reason is triggering `clippy::use_self`. Following it's suggestion results in code that is (to no surprise) uncompilable as it suggests replacing the enum name with Self.

### Lint Name

clippy::use_self

### Reproducer

I tried this code:

```rust
// Architecture
/// Represents a system architecture
//
#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Hash, Clone)]
pub enum Architecture {
    #[allow(non_camel_case_types)]
    amd64,
    #[allow(non_camel_case_types)]
    arm64,
    #[allow(non_camel_case_types)]
    armv7,
}

```

I saw this happen:

```
warning: unnecessary structure name repetition
  --> libmangrove/src/platform.rs:10:10
   |
10 | pub enum Architecture {
   |          ^^^^^^^^^^^^ help: use the applicable keyword: `Self`
   |
   = note: `#[warn(clippy::use_self)]` implied by `#[warn(clippy::nursery)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self
```
I have no idea if I am causing this or if it is an actual bug as my codebase is quite large, but I am pretty sure this is a bug.

### Version

```text
rustc 1.65.0                                                                                                                                     
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by c0repwn3r** [comment]
It looks like this is a duplicate of #6902 which has been closed as this is a very similar situation. Either a bug with serde or here.

**Comment by kraktus** [comment]
Hey, thanks for reporting,

I cannot reproduce this issue on stable or nightly: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e78b9597b9e6ab6d18be4526180cd8d1

Updating your rust version will fix it.

**Comment by kraktus** [comment]
Blast from the past! 😄
